lib/elixir/lib/json.ex

```diff
@@ -0,0 +1,281 @@
+    end
+  end
+end
+
+defimpl JSON.Encoder, for: BitString do
+  def encode(value, _encoder) do
+    :elixir_json.encode_binary(value)
+  end
+end
+
+defimpl JSON.Encoder, for: List do
+  def encode(value, encoder) do
+    :elixir_json.encode_list(value, encoder)
+  end
+end
+
+defimpl JSON.Encoder, for: Integer do
+  def encode(value, _encoder) do
+    :elixir_json.encode_integer(value)
+  end
+end
+
+defimpl JSON.Encoder, for: Float do
+  def encode(value, _encoder) do
+    :elixir_json.encode_float(value)
+  end
+end
+
+defimpl JSON.Encoder, for: Map do
+  def encode(value, encoder) do
+    :elixir_json.encode_map(value, encoder)
+  end
+end
+
+defmodule JSON do
+  @moduledoc ~S"""
+  JSON encoding and decoding.
+
+  Both encoder and decoder fully conform to [RFC 8259](https://tools.ietf.org/html/rfc8259) and
+  [ECMA 404](https://ecma-international.org/publications-and-standards/standards/ecma-404/)
+  standards.
+
+  ## Encoding
+
+  Elixir built-in data structures are encoded to JSON as follows:
+
+  | **Elixir**             | **JSON** |
+  |------------------------|----------|
+  | `integer() \| float()` | Number   |
+  | `true \| false `       | Boolean  |
+  | `nil`                  | Null     |
+  | `binary()`             | String   |
+  | `atom()`               | String   |
+  | `list()`               | Array    |
+  | `%{binary() => _}`     | Object   |
+  | `%{atom() => _}`       | Object   |
+  | `%{integer() => _}`    | Object   |
+
+  You may also implement the `JSON.Encoder` protocol for custom data structures.
+
+  ## Decoding
+
+  Elixir built-in data structures are decoded from JSON as follows:
+
+  | **JSON** | **Elixir**             |
+  |----------|------------------------|
+  | Number   | `integer() \| float()` |
+  | Boolean  | `true \| false`        |
+  | Null     | `nil`                  |
+  | String   | `binary()`             |
+  | Object   | `%{binary() => _}`     |
+
+  """
+
+  @moduledoc since: "1.18.0"
+
+  @doc ~S"""
+  Encodes the given term to JSON as a binary.
+
+  The second argument is a function that is recursively
+  invoked to encode a term.
+
+  ## Examples
+
+      iex> JSON.encode([123, "string", %{key: "value"}])
+      "[123,\"string\",{\"key\":\"value\"}]"
+
+  """
+  def encode(term, encoder \\ &encode_value/2) do
+    IO.iodata_to_binary(encoder.(term, encoder))
+  end
+
+  @doc ~S"""
+  Encodes the given term to JSON as an iodata.
+
+  This is the most efficient format if the JSON is going to be
+  used for IO purposes.
+
+  The second argument is a function that is recursively
+  invoked to encode a term.
+
+  ## Examples
+
+      iex> data = JSON.encode_to_iodata([123, "string", %{key: "value"}])
+      iex> IO.iodata_to_binary(data)
+      "[123,\"string\",{\"key\":\"value\"}]"
+
+  """
+  def encode_to_iodata(term, encoder \\ &encode_value/2) do
+    encoder.(term, encoder)
+  end
+
+  @doc """
+  A shortcut for `encode/2` used for compatibility purposes.
```

Comment:
Actually, I'm not sure it makes sense to provide this one, given `encode/2` won't be returning an `:ok` tuple, it won't be a drop-in replacement anyway.

People will have to migrate:
- `Jason.encode!` to `JSON.encode`
- `Jason.encode` to `JSON.encode` after removing the tuple-matching

The fact that `JSON.encode!` is the same as `JSON.encode` and not a raising alternative is also confusing.

Reply:
It is only meant to be temporary because Phoenix and most projects only use the bang function. This would allow someone to use Phoenix with JSON.

Reply:
I see. For this use case, could we consider something like declaring a

```elixir
defmodule JSON.Compat do
  def encode_to_iodata!(data), do: JSON.encode_to_iodata(data)
  ...
end
```

and use `JSON.Compat` as the pluggable backend? This would have just the necessary bang APIs, without exposing options. It doesn't even have to be defined in Elixir, could be done in Phoenix.

Reply:
We would need to define this module for Postgrex, EctoSQL, Phoenix, and probably many others. The goal of this function was to provide an easier onboarding ramp.

Reply:
Then in the future, they can depend specifically on JSON, and then we can deprecate this. But for a while they will need to support both Jason and JSON.

Reply:
It could be in Elixir too in this case 🙂 But it would solve several issues:
- no confusion due to the bang being the same as non-bang
- clear objective: "compat" is in the name
- does not pollute the regular `JSON` module, people won't be tempted to call it directly

Reply:
Yes, you are right. A separate module may be better but now I am wondering if it is worth it.